File: python/camera_ui_sdk/sensor/leak.py
from __future__ import annotations

from collections.abc import Mapping
from enum import StrEnum
from typing import Any, Generic, Literal, Protocol, overload, runtime_checkable

from typing_extensions import TypedDict, TypeVar

from ..observable import Observable
from .base import Sensor, SensorCategory, SensorLike, SensorType


class LeakProperty(StrEnum):
    """Property names of a leak sensor."""

    Detected = "detected"
    """Whether a leak is detected."""


class LeakSensorProperties(TypedDict):
    """Property values of a leak sensor."""

    detected: bool


class LeakPropertyChangeData(TypedDict):
    """Property change payload emitted on LeakSensorLike.onPropertyChanged."""

    property: str
    """Name of the changed property, a LeakProperty value."""
    value: bool
    """New value of the property."""


TStorage = TypeVar("TStorage", bound=Mapping[str, Any], default=dict[str, Any])


@runtime_checkable
class LeakSensorLike(SensorLike, Protocol):
    """Read-only proxy interface for a leak sensor."""

    @property
    def type(self) -> SensorType:
        return SensorType.Leak

    @property
    def onPropertyChanged(self) -> Observable[LeakPropertyChangeData]: ...

    @overload
    def getValue(self, property: Literal[LeakProperty.Detected]) -> bool | None: ...
    @overload
    def getValue(self, property: str) -> object | None: ...


class LeakSensor(Sensor[LeakSensorProperties, TStorage, str], Generic[TStorage]):
    """Water leak detector sensor."""

    _requires_frames = False

    def __init__(self, name: str = "Leak Sensor", *, native_id: str | None = None) -> None:
        super().__init__(name, native_id=native_id)
        self._write_state({LeakProperty.Detected.value: False})

    @property
    def type(self) -> SensorType:
        return SensorType.Leak

    @property
    def category(self) -> SensorCategory:
        return SensorCategory.Sensor

    @property
    def detected(self) -> bool:
        return bool(self.props.detected)

    def setDetected(self, value: bool) -> None:
        """Report leak detection state.

        Args:
            value: True when a water leak is currently detected.

        Example:
            ```python
            leak.setDetected(True)
            ```
        """
        self._write_state({LeakProperty.Detected.value: value})

    async def updateValue(self, property: str, value: Any) -> None:
        """Read-only sensor: external writes are ignored."""

File: node/src/sensor/leak.ts
import { Sensor, SensorType, SensorCategory } from './base.js';
import { defineSensor, SensorDomain } from './meta.js';

import type { Observable } from '../observable/index.js';
import type { PropertyChangeOf, SensorLike, SensorOptions } from './base.js';

/**
 * Property names of a leak sensor.
 *
 * @internal
 */
export enum LeakProperty {
  /** Whether a leak is detected. */
  Detected = 'detected',
}

/**
 * Property values of a leak sensor.
 *
 * @internal
 */
export interface LeakSensorProperties {
  [LeakProperty.Detected]: boolean;
}

/** Read-only proxy interface for a leak sensor. */
export interface LeakSensorLike extends SensorLike {
  readonly type: SensorType.Leak;
  readonly onPropertyChanged: Observable<PropertyChangeOf<LeakSensorProperties>>;

  getValue(property: LeakProperty.Detected): boolean | undefined;
  getValue(property: string): unknown;
}

/** Water leak detector sensor. */
export class LeakSensor<TStorage extends object = Record<string, any>> extends Sensor<LeakSensorProperties, TStorage> {
  readonly type = SensorType.Leak;
  readonly category = SensorCategory.Sensor;

  constructor(name = 'Leak Sensor', options?: SensorOptions) {
    super(name, options);

    this._writeState({ [LeakProperty.Detected]: false });
  }

  get detected(): boolean {
    return this.props.detected;
  }

  /**
   * Report leak detection state.
   *
   * @param value - True when a leak is currently detected.
   *
   * @example
   * ```ts
   * leak.setDetected(true);
   * ```
   */
  setDetected(value: boolean): void {
    this._writeState({ [LeakProperty.Detected]: value });
  }

  /**
   * Read-only sensor: external writes are ignored.
   *
   * @internal
   */
  updateValue(_property: string, _value: unknown): void {}
}

/** Registry metadata for {@link LeakSensor}. */
export const leakMeta = defineSensor({
  type: SensorType.Leak,
  category: SensorCategory.Sensor,
  assignmentKey: 'leak',
  multiProvider: true,
  isDetectionType: false,
  properties: {
    [LeakProperty.Detected]: { type: 'boolean', writable: true },
  },
  shortcutable: true,
  cascadeTrigger: { property: LeakProperty.Detected, value: true, sustained: true },
  virtual: { properties: { [LeakProperty.Detected]: false } },
  semantics: {
    domain: SensorDomain.Binary,
    stateProperty: LeakProperty.Detected,
    commandProperty: LeakProperty.Detected,
    deviceClass: 'moisture',
  },
});
